### DISCUSSION TITLE
Global Static Data

### DISCUSSION THREAD

**Discussion by nathnhughes** [vercel/next.js#37043] [Help]
I've wrestled with this topic a few times and am keen to see if the community can provide any insights or potential solutions. 

Often in projects I like to use some type of `<Layout/>` component, usually encompassing the `<Header/>` and `<Footer/>`, that wraps the page content and then persists across route changes, helping to give that SPA feel and preventing re-rending the same components for every page. 

In Next, this means including it in the `_app.tsx` file like so: 

```jsx
/* Layout.tsx */
interface Props {}

const Layout: React.FC<PropsWithChildren<Props>> = ({ children }) => (
  <>
    <Header />
      {children}
    <Footer />
  </>
)

export default Layout
```
```jsx
/* _app.tsx */

const App = ({ Component, pageProps }: AppProps): JSX.Element => (
  <Layout />
    <Component {...pageProps} />
  </Layout>
)

export default App
```

However, the issue arises when we need to get data into these components. 

On a recent project, we had a rather elaborate mega-menu contained within the header. The data for this was stored in a CMS but was pretty much static _(potentially changing every 3-6 months)_ so it didn't make sense to re-fetch it for every user. 

Here is where the problem lies...

This data was essentially static, or static enough that it could be fetched at build time, so ideally we'd fetch it once then and use the data to statically generate the HTML for the menu component. However, as you can only fetch static props at the _page_ level, there's currently not an efficient way to do this in Next _(that I'm aware of)_.

<br>

## Potential Solutions
Here are some options I've considered in the past:

- **Fetch the data in every pages GSP call:** This is less than ideal as we're then having to add the `<Header/>` to every page and not adding it in the _app.tsx file.
- **Fetch the data in the App using getInitialProps:** I'm less keen on this as GIP will run on a per-user basis and I'd rather not have to continually hit the CMS's API to re-fetch the same data
- **Storing the static JSON files in the bundle:** Also an option but again not ideal as it requires manually updating the data by hand if/when something changes

Would love to know if anyone else has faced a similar issue and how they've handled it. Would also be great to hear if Next plans to add any features to address this like using `gSP` in the App component or component-level data fetching.

<br>

## Linked issues
A few issues i've found discussing similar topics: 

- https://github.com/vercel/next.js/discussions/26346
- https://github.com/vercel/next.js/discussions/36806
- https://github.com/vercel/next.js/discussions/12595

**Comment by icyJoseph** [comment]
**Not a solution, just an observation**

> Fetch the data in every pages GSP call: This is less than ideal as we're then having to add the <Header/> to every page and not adding it in the _app.tsx file.

If fetching the data with GSP on every page, at build time, is not such a big deal for you, but rather, the problem is that you'd import and use the Layout on every page, then do remember that `pageProps`, in `_app.tsx`, contains the results of GSP for each page, that implements GSP.

> pageProps is an object with the initial props that were preloaded for your page by one of our [data fetching methods](https://nextjs.org/docs/basic-features/data-fetching/overview), otherwise it's an empty object.

```tsx
import type { NextPage } from "next";

const fetchData = () => {
  return new Promise<{ foo: string }>((resolve) => {
    setTimeout(resolve, 200, { foo: "bar" });
  });
};

const Home: NextPage = () => {
  return <div>Home</div>;
};

export const getStaticProps = async () => {
  const data = await fetchData();
  return { props: data }; 
};

export default Home;
```

Then in `_app.tsx`:

```tsx
import type { AppProps } from "next/app";
import Layout from "../components/Layout";

function MyApp({ Component, pageProps }: AppProps) {
  const { foo } = pageProps;

  return (
    <Layout foo={foo}>
      <Component {...pageProps} />
    </Layout>
  );
}

export default MyApp;
```

with Layout:

```tsx
const Layout = ({ foo, children }: { foo: string; children: ReactNode }) => (
  <>
    <header>{foo}</header>

    {children}

    <footer />
  </>
);

export default Layout;
```

Still remembering to include GSP, or the call to fetch this data, on every page, which, is not super ideal, and in some cases, _error pages would have to also implement this, IIRC. 

I would also love to hear other attempts at this pattern. I'll review some of my notes tomorrow.

**Reply by nathnhughes** [reply]
Interesting, I'd never thought about it like that before! 

Like you say, it's not ideal to have to include your fetch function in every page's` gSP` call, however, it's certainly less overhead that rendering the component on every page.

The only other immediate issue that springs to mind is using this on pages that are already fetching a lot of static data.

I recently worked on a site that fetched articles from a Contentful CMS. The articles were written with rich text and the way Contentful formats this is in the form of an _(often gigantic)_ JSON object. 

Seeing as Next has to provide these props on the server and rehydrate the static markup on the client, this rich text JSON would be included in the `__NEXT_DATA__` script and we'd regularly get [`large-page-data`](https://nextjs.org/docs/messages/large-page-data) logs from the server reporting page data of 150kb+.

So in that situation, this may not be an option, however, that is somewhat of a fringe case.
